src/rubrix/labeling/text_classification/weak_labels.py

```diff
@@ -770,6 +746,63 @@ def change_mapping(self, label2int: Dict[str, int]):
         self._label2int = label2int.copy()
         self._int2label = {val: key for key, val in self._label2int.items()}
 
+    def extend_matrix(
+        self,
+        thresholds: Union[List[float], np.ndarray],
+        embeddings: Optional[np.ndarray] = None,
+        gpu: bool = False,
+    ):
+        """Extends the weak label matrix through embeddings according to the similarity thresholds for each rule.
+           Implementation based on `Epoxy <https://github.com/HazyResearch/epoxy>`__.
+
+        Args:
+            thresholds: An array of thresholds between 0.0 and 1.0, one for each column of the weak labels matrix.
+                Each one stands for the minimum cosine similarity between two sentences for a rule to be extended.
+            embeddings: Embeddings for each row of the weak label matrix.
+                If not provided, we will use the ones from the last ``WeakLabels.extend_matrix()`` call.
+            gpu: If True, perform FAISS similarity queries on GPU.
+
+        Examples:
+            >>> # Choose any model to generate the embeddings.
+            >>> from sentence_transformers import SentenceTransformer
+            >>> model = SentenceTransformer('all-mpnet-base-v2', device='cuda')
+            >>>
+            >>> # Generate the embeddings and set the thresholds.
+            >>> weak_labels = WeakLabels(dataset="my_dataset")
+            >>> embeddings = np.array([ model.encode(rec.text) for rec in weak_labels.records() ])
+            >>> thresholds = [0.6] * len(weak_labels.rules)
+            >>>
+            >>> # Extend the weak labels matrix.
+            >>> weak_labels.extend_matrix(thresholds, embeddings)
+            >>>
+            >>> # Calling the method below will now retrieve the extended matrix.
+            >>> weak_labels.matrix()
+            >>>
+            >>> # Subsequent calls without the embeddings parameter will reutilize the faiss index built on the first call.
+            >>> thresholds = [0.75] * len(weak_labels.rules)
+            >>> weak_labels.extend_matrix(thresholds)
+            >>> weak_labels.matrix()
+        """
+        support = [
+            np.argwhere(self._matrix[:, i] != self._label2int[None]).flatten()
+            for i in range(self._matrix.shape[1])
+        ]
+
+        abstains = [
+            np.argwhere(self._matrix[:, i] == self._label2int[None]).flatten()
+            for i in range(self._matrix.shape[1])
+        ]
+
+        self._extended_matrix = self._extend_matrix(
+            thresholds, embeddings, abstains, support, gpu
+        )
+
+        # keep the original weak label row, for which at least on rule did not abstain
+        recs_with_votes = np.argwhere(
+            (self._matrix != self._label2int[None]).sum(-1) > 0
+        ).flatten()
+        self._extended_matrix[recs_with_votes] = self._matrix[recs_with_votes]
```

Comment:
Should we apply this post-process to the multilabel wl matrix?

Reply:
Hm, no, for the multi-label case this does not make much sense, since you do not create conflicts between labels by extending the rules. For example:

text = "This is a sad story and it makes me angry"

**single-label case**
Imagine a sentiment analysis (negative, positive) and you have a rule `if "sad" in text: return "negative"`; If you extend the matrix you probably want to avoid that a "positive" rule extends to this example since you already have an explicit rule that votes for this example (you basically give the explicit rules more weight).

**multi-label case**
Imaging a emotion detection task (sad, angry) and you have a rule `if "sad" in text: return "sad"`; If you extend the matrix you don't mind if an "angry" rule also extends to this example;

Reply:
mkay